File: Allocate minimum number of pages - GFG/allocate-minimum-number-of-pages.cpp
//{ Driver Code Starts
// Initial template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template in C++

class Solution 
{    bool isValid(int A[],int N,int M,int m){
     int student=1,sum=0;
     for(int i=0;i<N;i++){
         sum+=A[i];
         if(sum>m){
             student++;
             sum=A[i];
             if(student>M)return false;
         }
     }
     return true;
     }
    public:
    //Function to find minimum number of pages.
    int findPages(int A[], int N, int M) 
    {
        int s=*max_element(A,A+N);
        int e=accumulate(A,A+N,0);
        int soln=-1;
        if(M>N)return -1;
        while(s<=e){
            int m=s+(e-s)/2;
            if(isValid(A,N,M,m)){
                soln=m;
                e=m-1;
            }
            else{
                s=m+1;
            }
        }
        return soln;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        int A[n];
        for(int i=0;i<n;i++){
            cin>>A[i];
        }
        int m;
        cin>>m;
        Solution ob;
        cout << ob.findPages(A, n, m) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Allocate minimum number of pages - GFG/allocate-minimum-number-of-pages.java
//{ Driver Code Starts
//Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;
class GFG {
	public static void main (String[] args) {
		Scanner sc=new Scanner(System.in);
		
		int t=sc.nextInt();
		
		while(t-->0)
		{
		    int n=sc.nextInt();
		    int a[]=new int[n];
		    
		    for(int i=0;i<n;i++)
		    {
		        a[i]=sc.nextInt();
		    }
		    int m=sc.nextInt();
		    Solution ob = new Solution();
		    System.out.println(ob.findPages(a,n,m));
		}
	}
}
// } Driver Code Ends


//User function Template for Java

class Solution 
{
     public static int helper(int[] A,int k){
        int count=1;
        int prefix=0;
        for(int i:A){
            prefix+=i;
            if(prefix>k){
                prefix=i;
                count++;
            }
        }
        return count;
    }
    public static int findPages(int[]A,int N,int M)
    {   if(M>N)return -1;
        int min=Integer.MIN_VALUE;
        int max=0,ans=-1;
        for(int i:A){
            min=Math.max(i,min);
            max+=i;
        }
        while(min<=max){
            int mid=(min+max)/2;
            int students=helper(A,mid);
            if(M<students){
                min=mid+1;
            }
            else{
                ans=mid;
                max=mid-1;
            }
        }
        return ans;
    }
};